metadata-ingestion/src/datahub/ingestion/source/qlik_sense/qlik_api.py

```diff
@@ -0,0 +1,196 @@
+        """
+        socket_connection.send(json.dumps(request))
+        resp = socket_connection.recv()
+        if request["handle"] == -1:
+            resp = socket_connection.recv()
+        return json.loads(resp)
+
+    def _get_websocket_request_dict(
+        self, id: int, handle: int, method: str, params: Dict = {}
+    ) -> Dict:
+        return {
+            "jsonrpc": "2.0",
+            "id": id,
+            "handle": handle,
+            "method": method,
+            "params": params,
+        }
+
+    def get_spaces(self) -> List[Space]:
+        spaces: List[Space] = []
+        try:
+            response = self.session.get(f"{self.rest_api_url}/spaces")
+            response.raise_for_status()
+            for space_dict in response.json()[Constant.DATA]:
+                spaces.append(Space.parse_obj(space_dict))
+        except Exception:
+            self._log_http_error(message="Unable to fetch spaces")
+        return spaces
+
+    def _get_dataset(self, dataset_id: str) -> Optional[QlikDataset]:
+        try:
+            response = self.session.get(f"{self.rest_api_url}/data-sets/{dataset_id}")
+            response.raise_for_status()
+            return QlikDataset.parse_obj(response.json())
+        except Exception:
+            self._log_http_error(
+                message=f"Unable to fetch dataset with id {dataset_id}"
+            )
+        return None
+
+    def get_items(self) -> List[Item]:
+        items: List[Item] = []
+        try:
+            response = self.session.get(f"{self.rest_api_url}/items")
+            response.raise_for_status()
+            data = response.json()[Constant.DATA]
+            for item in data:
+                resource_type = item[Constant.RESOURCETYPE]
+                resource_attributes = item[Constant.RESOURCEATTRIBUTES]
+                if resource_type == Constant.APP:
+                    app = App.parse_obj(resource_attributes)
+                    app.sheets = self._get_app_sheets(app_id=app.id)
+                    items.append(app)
+                elif resource_type == Constant.DATASET:
+                    dataset = self._get_dataset(dataset_id=item[Constant.RESOURCEID])
+                    if dataset:
+                        items.append(dataset)
+
+        except Exception:
+            self._log_http_error(message="Unable to fetch items")
+        return items
+
+    def _get_sheet(
+        self,
+        socket_connection: WebSocket,
+        request_id: int,
+        current_handle: int,
+        sheet_id: str,
+    ) -> Optional[Sheet]:
+        try:
+            response = self._websocket_send_request(
+                socket_connection=socket_connection,
+                request=self._get_websocket_request_dict(
+                    id=request_id,
+                    handle=current_handle,
+                    method="GetObject",
+                    params={"qId": sheet_id},
+                ),
+            )
+            request_id += 1
```

Comment:
the bookkeeping related to `request_id` and handles should be contained within the _websocket_send_request or _get_websocket_request_dict methods

it might be worthwhile to create a `WebsocketConnection` class that handles everything about the request_ids and handles? I would like for the logic of talking with the Qlik API to be cleanly separated from the parsing logic